File: talor/src/tool/builtin/grep.py
"""Grep Tool for Talor.

This module provides the grep tool for searching file content.
"""

from __future__ import annotations

import logging
import re
from pathlib import Path

from pydantic import BaseModel, Field

from src.tool import Tool, ToolContext, ToolOutput


logger = logging.getLogger(__name__)


class GrepParams(BaseModel):
    """Parameters for grep tool."""

    pattern: str = Field(
        description="Regex pattern to search for"
    )
    path: str | None = Field(
        default=None,
        description="Path to search in (file or directory, default: workspace root)"
    )
    include: str | None = Field(
        default=None,
        description="Glob pattern for files to include (e.g., '*.py')"
    )
    case_sensitive: bool = Field(
        default=False,
        description="Whether the search is case-sensitive"
    )


async def grep_execute(params: GrepParams, ctx: ToolContext) -> ToolOutput:
    """Execute the grep tool.

    Args:
        params: Grep parameters
        ctx: Tool execution context

    Returns:
        ToolOutput with search results
    """
    # Determine search path
    if params.path:
        search_path = Path(params.path)
        if not search_path.is_absolute():
            search_path = ctx.worktree / search_path
    else:
        search_path = ctx.worktree

    # Validate path
    if not search_path.exists():
        return ToolOutput.error(
            f"Path not found: {params.path or '.'}",
            title="Grep Error"
        )

    try:
        # Compile regex
        flags = 0 if params.case_sensitive else re.IGNORECASE
        try:
            regex = re.compile(params.pattern, flags)
        except re.error as e:
            return ToolOutput.error(
                f"Invalid regex pattern: {e}",
                title="Grep Error"
            )

        # Collect files to search
        files_to_search = []

        if search_path.is_file():
            files_to_search.append(search_path)
        else:
            # Get files matching include pattern
            if params.include:
                files_to_search = list(search_path.rglob(params.include))
            else:
                # Default: search common text files
                for ext in ["*.py", "*.js", "*.ts", "*.jsx", "*.tsx", "*.md", "*.txt", "*.json", "*.yaml", "*.yml", "*.toml", "*.html", "*.css", "*.sh", "*.bash"]:
                    files_to_search.extend(search_path.rglob(ext))

        # Search files
        results = []
        max_results = 100
        max_context_lines = 2

        for file_path in files_to_search:
            if len(results) >= max_results:
                break

            if not file_path.is_file():
                continue

            try:
                content = file_path.read_text(encoding="utf-8")
                lines = content.splitlines()

                for i, line in enumerate(lines):
                    if len(results) >= max_results:
                        break

                    if regex.search(line):
                        # Get context lines
                        start = max(0, i - max_context_lines)
                        end = min(len(lines), i + max_context_lines + 1)

                        context = []
                        for j in range(start, end):
                            prefix = ">" if j == i else " "
                            context.append(f"{j + 1:4d}{prefix}| {lines[j]}")

                        try:
                            rel_path = file_path.relative_to(ctx.worktree)
                        except ValueError:
                            rel_path = file_path

                        results.append({
                            "file": str(rel_path),
                            "line": i + 1,
                            "context": "\n".join(context),
                        })

            except (UnicodeDecodeError, PermissionError):
                continue

        # Format output
        if not results:
            return ToolOutput(
                title=f"Grep '{params.pattern}' (no matches)",
                output="No matches found",
                metadata={"pattern": params.pattern, "matches_count": 0},
            )

        output_parts = []
        for result in results:
            output_parts.append(f"--- {result['file']}:{result['line']} ---")
            output_parts.append(result["context"])
            output_parts.append("")

        output = "\n".join(output_parts)

        truncated = len(results) >= max_results
        if truncated:
            output += f"\n(truncated, showing first {max_results} matches)"

        # Build metadata
        metadata = {
            "pattern": params.pattern,
            "matches_count": len(results),
            "files_searched": len(files_to_search),
            "truncated": truncated,
        }

        return ToolOutput(
            title=f"Grep '{params.pattern}' ({len(results)} matches)",
            output=output,
            metadata=metadata,
        )

    except Exception as e:
        logger.error(f"Error in grep: {e}")
        return ToolOutput.error(
            f"Error in grep: {e}",
            title="Grep Error"
        )


# Define the grep tool
GrepTool = Tool.define(
    id="grep",
    description="""Search for a pattern in files.

Use this tool to search file content using regex patterns.
Results include context lines around each match.

IMPORTANT:
- Use regex syntax for patterns
- Escape special characters: . * + ? ^ $ [ ] { } | \\
- Default is case-insensitive

Examples:
- Find function: {"pattern": "def my_function"}
- Find imports: {"pattern": "^import", "include": "*.py"}
- Find TODO: {"pattern": "TODO|FIXME", "path": "src"}
- Case-sensitive: {"pattern": "MyClass", "case_sensitive": true}
""",
    parameters=GrepParams,
    execute=grep_execute,
)

File: src/main/tools/builtin/grep.ts
import { readdirSync, readFileSync, existsSync, statSync, realpathSync } from 'fs'
import { join } from 'path'
import { toolRegistry } from '../registry'
import type { ToolExecuteContext, ValidationResult } from '../types'
import { resolveToolPath, isPathSensitive } from '../path-guard'

const SKIP_DIRS = new Set(['node_modules', '.git', '.cache'])
const MAX_RESULTS = 100
const CONTEXT_LINES = 2

const REDOS_PATTERNS = [
  /\(\?[^)]*\)\+/,
  /\([^)]*\+[^)]*\)\+/,
  /\([^)]*\+[^)]*\)\*/,
  /\([^)]*\*[^)]*\)\+/,
  /\([^)]*\*[^)]*\)\*/,
]

function isSuspectedReDoS(pattern: string): boolean {
  return REDOS_PATTERNS.some(re => re.test(pattern))
}

const grepTool = {
  name: 'grep',
  description:
    'Search for a pattern in files using regex. Returns matching lines with context. ' +
    'Never use to locate skill definitions — skills live in memory, use the `skill` tool.',
  parameters: {
    type: 'object',
    properties: {
      pattern: { type: 'string', description: 'Regex pattern to search for' },
      path: { type: 'string', description: 'File or directory path to search in (default: workspace root)', default: '.' },
      include: { type: 'string', description: 'Glob pattern for files to include (e.g., "*.ts")' },
      caseSensitive: { type: 'boolean', description: 'Whether search is case-sensitive', default: false },
    },
    required: ['pattern'],
  },

  validate(input: unknown): ValidationResult {
    const params = input as { pattern?: unknown }
    if (typeof params.pattern !== 'string' || !params.pattern.trim())
      return { ok: false, error: 'Missing required parameter: "pattern" must be a non-empty string.' }
    try {
      new RegExp(params.pattern)
    } catch (e) {
      return { ok: false, error: `Invalid regex pattern: ${e instanceof Error ? e.message : String(e)}` }
    }
    return { ok: true }
  },

  async execute(input: unknown, context: ToolExecuteContext): Promise<{ output: unknown }> {
    const { workspace } = context
    const params = input as { pattern: string; path?: string; include?: string; caseSensitive?: boolean }

    if (!workspace) {
      return { output: 'Workspace not set. Please set workspace first.' }
    }

    if (isPathSensitive(workspace)) {
      return { output: 'Cannot search sensitive system directory' }
    }

    const targetPath = params.path || '.'
    const guard = resolveToolPath(targetPath, workspace)
    if (guard.status === 'sensitive') {
      return { output: 'Cannot access sensitive system path' }
    }
    if (guard.status === 'needs_consent') {
      const approved = await context.requestPermission?.({
        toolName: 'grep',
        reason: 'path_outside_workspace',
        absPath: guard.absPath,
        inputSummary: `${params.pattern} in ${targetPath}`,
      })
      if (!approved) return { output: 'Cannot access path outside workspace (user denied).' }
    }
    const resolvedPath = guard.absPath

    if (!existsSync(resolvedPath)) {
      return { output: `Path not found: ${targetPath}` }
    }

    try {
      if (isSuspectedReDoS(params.pattern)) {
        return { output: 'Pattern rejected: potential ReDoS risk (nested quantifiers detected)' }
      }

      let regex: RegExp
      try {
        const flags = params.caseSensitive ? 'g' : 'gi'
        regex = new RegExp(params.pattern, flags)
      } catch (e) {
        return { output: `Invalid regex pattern: ${e instanceof Error ? e.message : String(e)}` }
      }

      const filesToSearch: string[] = []
      const stats = statSync(resolvedPath)

      if (stats.isFile()) {
        filesToSearch.push(resolvedPath)
      } else {
        const includePattern = params.include || '*.{js,ts,jsx,tsx,py,md,txt,json,yaml,yml,toml,html,css,sh}'
        const exts: string[] = []
        
        const addExt = (pattern: string) => {
          const starIdx = pattern.indexOf('*.')
          if (starIdx !== -1) {
            exts.push(pattern.substring(starIdx + 1))
          }
        }
        
        if (includePattern.includes('{') && includePattern.includes('}')) {
          const braceMatch = includePattern.match(/\{([^}]+)\}/)
          if (braceMatch) {
            braceMatch[1].split(',').forEach(e => addExt(e.trim()))
          }
        } else if (includePattern.includes('*')) {
          addExt(includePattern)
        }
        
        if (exts.length === 0) exts.push('*')

        const matchAllExts = exts.includes('*')

        let realWorkspace: string
        try {
          realWorkspace = realpathSync(workspace)
        } catch {
          realWorkspace = workspace
        }

        function collectFiles(dir: string, depth: number): void {
          if (depth > 5 || filesToSearch.length >= MAX_RESULTS) return
          try {
            const entries = readdirSync(dir, { withFileTypes: true })
            for (const entry of entries) {
              if (filesToSearch.length >= MAX_RESULTS) break

              if (entry.isDirectory()) {
                if (!SKIP_DIRS.has(entry.name)) collectFiles(join(dir, entry.name), depth + 1)
                continue
              }

              if (!entry.isFile()) continue

              const ext = entry.name.includes('.') ? '.' + entry.name.split('.').pop() : ''
              if (!matchAllExts && !exts.includes(`*.${ext}`) && !exts.some(e => e === ext)) continue

              const candidatePath = join(dir, entry.name)
              try {
                const realCandidate = realpathSync(candidatePath)
                if (realCandidate.startsWith(realWorkspace)) filesToSearch.push(candidatePath)
              } catch {
                // skip unresolvable symlinks
              }
            }
          } catch {
            // skip inaccessible
          }
        }

        collectFiles(resolvedPath, 0)
      }

      const results: string[] = []

      for (const filePath of filesToSearch) {
        if (results.length >= MAX_RESULTS) break

        try {
          const content = readFileSync(filePath, 'utf-8')
          const lines = content.split('\n')

          for (let i = 0; i < lines.length; i++) {
            if (results.length >= MAX_RESULTS) break

            regex.lastIndex = 0
            if (regex.test(lines[i])) {
              const start = Math.max(0, i - CONTEXT_LINES)
              const end = Math.min(lines.length, i + CONTEXT_LINES + 1)
              const context: string[] = []

              for (let j = start; j < end; j++) {
                const prefix = j === i ? '>' : ' '
                context.push(`${String(j + 1).padStart(4)}${prefix}| ${lines[j]}`)
              }

              const relPath = filePath.replace(workspace, '').replace(/^\//, '')
              results.push(`--- ${relPath}:${i + 1} ---`)
              results.push(context.join('\n'))
              results.push('')
            }
          }
        } catch {
          // skip binary/inaccessible
        }
      }

      if (results.length === 0) {
        return { output: 'No matches found' }
      }

      let output = results.join('\n')
      if (results.length >= MAX_RESULTS) {
        output += `\n(truncated, showing first ${MAX_RESULTS} matches)`
      }

      return { output }
    } catch (err) {
      return { output: err instanceof Error ? err.message : String(err) }
    }
  },
}

export function registerBuiltinTools(): void {
  toolRegistry.register(grepTool)
}
